### PR TITLE
Adding integration test cases for OAuth DCRM implementation

### PR DESCRIPTION
**Pull request by Abilashini**

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth/dcrm/OAuthDCRMDeleteTestCase.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
+package org.wso2.identity.integration.test.oauth.dcrm;
+
```
Comment (isharak) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth/dcrm/OAuthDCRMDeleteTestCase.java]: licenses headers are missing

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth/dcrm/OAuthDCRMReadTestCase.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        HttpResponse response = client.execute(request);
+        Assert.assertNotNull(response, "Service Provider read request failed");
+
+        BufferedReader rd = new BufferedReader(new InputStreamReader(response.getEntity().getContent()));
+        Object obj = JSONValue.parse(rd);
+        Assert.assertNotNull(obj, "Returned response should have produced a valid JSON.");
+
+        JSONObject jsonObject = (JSONObject) obj;
+
+        Assert.assertEquals((String) jsonObject.get(OAuthDCRMConstants.CLIENT_NAME), serviceProvider.getClientName());
```
Comment (dimuthud) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth/dcrm/OAuthDCRMReadTestCase.java]: As a practice always put error description to assertEquals(..)

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth/dcrm/OAuthDCRMReadTestCase.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    @Test(alwaysRun = true, description = "Read request with an invalid client ID")
+    public void testReadServiceProviderWithInvalidClientID() throws IOException {
+        HttpGet request = new HttpGet(OAuthDCRMConstants.CLIENT_CONFIGURATION_ENDPOINT
+                + OAuthDCRMConstants.INVALID_CLIENT_ID);
+        request.addHeader(HttpHeaders.AUTHORIZATION, OAuthDCRMConstants.AUTHORIZATION);
+
+        HttpResponse response = client.execute(request);
+        Assert.assertNotNull(response, "Service Provider read request failed");
+
+        BufferedReader rd = new BufferedReader(new InputStreamReader(response.getEntity().getContent()));
```
Comment (dimuthud) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth/dcrm/OAuthDCRMReadTestCase.java]: Please close the buffered reader. rd.close();

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth/dcrm/OAuthDCRMDeleteTestCase.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    @Test(alwaysRun = true, description = "Delete Service Provider")
+    public void testDeleteServiceProvider() throws IOException {
+
+        HttpDelete request = new HttpDelete(OAuthDCRMConstants.CLIENT_CONFIGURATION_ENDPOINT
+                + serviceProvider.getClientID());
+        request.addHeader(HttpHeaders.AUTHORIZATION, OAuthDCRMConstants.AUTHORIZATION);
+
+        HttpResponse response = client.execute(request);
+        Assert.assertNotNull(response, "Service Provider Delete request failed");
```
Comment (dimuthud) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth/dcrm/OAuthDCRMDeleteTestCase.java]: You can use static imports here.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth/dcrm/OAuthDCRMReadTestCase.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    public void testInit() throws Exception {
+        client = new DefaultHttpClient();
+        serviceProviderRegister = new ServiceProviderRegister();
+
+        JSONObject object = new JSONObject();
+        object.put(OAuthDCRMConstants.CLIENT_NAME, "ReadApp");
+        object.put(OAuthDCRMConstants.GRANT_TYPES, "implicit");
+        object.put(OAuthDCRMConstants.REDIRECT_URIS, "http://ReadApp.com");
+
+        serviceProvider = serviceProviderRegister.register(object.toJSONString());
```
Comment (dimuthud) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth/dcrm/OAuthDCRMReadTestCase.java]: We dont need the SP to be remain the Server after the test is completed. isnt it? If so add a @AfterClass and remove the SP added after the test execution .

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth/dcrm/OAuthDCRMUpdateTestCase.java [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+import org.wso2.identity.integration.test.oauth.dcrm.bean.ServiceProvider;
+import org.wso2.identity.integration.test.oauth.dcrm.util.OAuthDCRMConstants;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+
+public class OAuthDCRMUpdateTestCase {
```
Comment (dimuthud) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oauth/dcrm/OAuthDCRMUpdateTestCase.java]: Use a test class description always to indicate what actually this test class does.

**Comment by JKAUSHALYA** [comment]
Merged with https://github.com/wso2/product-is/pull/1464

**Pull request closed without merging**